### ISSUE TITLE
[BUG] ImageURL does not appear in legends.

### ISSUE THREAD

**Issue by ethankruse** [bokeh/bokeh#12824]
### Software versions

Python version      :  3.10.9 (main, Jan 11 2023, 09:18:20) [Clang 14.0.6 ]
IPython version     :  8.10.0
Tornado version     :  6.2
Bokeh version       :  3.0.3
BokehJS static path :  /Users/ekruse/anaconda/envs/py310/lib/python3.10/site-packages/bokeh/server/static
node.js version     :  (not installed)
npm version         :  (not installed)
Operating system    :  macOS-10.16-x86_64-i386-64bit

### Expected behavior

I'm not entirely sure if this is a bug or a feature request, but I'd like some clarification or additional documentation about this. I'd expect that the `ImageURL` image would appear in some form in the legends, since it seems to function identically to other things you can put in legends.

### Observed behavior

You can get the legend label in there, and you can even toggle it visible or not by clicking either on that label or the empty space where the icon is supposed to show up. It's just that nothing is there.

### Example code

```Python
import numpy as np

from bokeh.io import curdoc, show
from bokeh.models import ColumnDataSource, Grid, ImageURL, LinearAxis, Plot, Range1d, Legend, Line

url = "https://static.bokeh.org/logos/logo.png"
N = 5

source = ColumnDataSource(dict(
    url = [url]*5,
    x1  = np.linspace(  0, 150, N),
    y1  = np.linspace(  0, 150, N),
    w1  = np.linspace( 10,  50, N),
    h1  = np.linspace( 10,  50, N),
    x2  = np.linspace(-50, 150, N),
    y2  = np.linspace(  0, 200, N),
))

xdr = Range1d(start=-100, end=200)
ydr = Range1d(start=-100, end=200)

plot = Plot(
    title=None, x_range=xdr, y_range=ydr, width=500, height=500,
    min_border=0, toolbar_location=None)

x = np.linspace(-100, 200, 400)
y = np.linspace(-100, 200, 400)

source1 = ColumnDataSource(data=dict(x=x, y=y))
line_glyph = Line(x="x", y="y", line_color="navy", line_width=2, line_dash="dashed")
line = plot.add_glyph(source1, line_glyph)

image1 = ImageURL(url="url", x="x1", y="y1", w="w1", h="h1", anchor="center")
glyph1 = plot.add_glyph(source, image1)

image2 = ImageURL(url="url", x="x2", y="y2", w=20, h=20, anchor="top_left")
glyph2 = plot.add_glyph(source, image2)

xaxis = LinearAxis()
plot.add_layout(xaxis, 'below')

yaxis = LinearAxis()
plot.add_layout(yaxis,'left')

plot.add_layout(Grid(dimension=0, ticker=xaxis.ticker))
plot.add_layout(Grid(dimension=1, ticker=yaxis.ticker))

legend = Legend(items=[("line", [line]), ("image 1", [glyph1]),
                       ("image 2", [glyph2])], location="top")
plot.add_layout(legend, 'above')
plot.legend.click_policy="hide"

curdoc().add_root(plot)

show(plot)
```

**Comment by bryevdv** [comment]
With a line or marker, or even a filled shape, it's generally clear what constitutes a "representative" sample to put a legend. But for images this is not the case. Certainly it's possible for images to be small, so that the entire image could be shown  in the legend. But that's not true in general&mdash;images might be hundreds or thousands of pixels in size. Until/unless someone suggests a reasonable API or scheme that can be do something sensible for all images, consistently, regardless of their size, I would be 👎  on doing anything here besides document the situation (have a feature that works OK on some images but not others is worse than just not supporting images in legends at all IMO)

What do other plotting libraries do with respect to images and legends? @bokeh/dev

**Comment by ethankruse** [comment]
Yeah that's understandable. For reference, my [use case](https://ethankruse.github.io/exoplots/) is currently using `ImageURL` as a stand-in for custom markers in a scatterplot, and I'd like that to show up in the legend (I want the Solar System group to show some symbol or image in the legend next to the label). The images are small (100x100 right now). Maybe there's another way to go about it that doesn't involve `ImageURL`.

I actually thought this was easy in matplotlib, but apparently not. They do [offer](https://matplotlib.org/stable/gallery/lines_bars_and_markers/marker_reference.html) quite a bit more flexibility out of the box than currently in bokeh. And it looks like with a little [work](https://petercbsmith.github.io/marker-tutorial.html) I could get what I want done in matplotlib. All of that looks like quite a big bit of development work for bokeh though, so I get not wanting that done.

In the meantime, is there any other possible workaround? What is different between the `Line` and `ImageURL`? On the surface it looks like in both cases I'm feeding the legend a `GlyphRenderer` object. In any case, is there any way to manually edit the glyph that goes next to the legend label and put something else there? Or include an image as part of the legend label or something?

**Comment by bryevdv** [comment]
>  using ImageURL as a stand-in for custom markers 

I suspected that might be the case. A possible avenue to explore in the future, that would keep API and expectations clear, would be an `ImageMarker` that draws a centered image, and also just draws the entire image scaled into the legend. The coordinates would be vectorized, but (unlike other image glyphs) the image would not be vectorized, which means it is plain what the representative to draw is. The could behave well under the reasonable assumption that scatter markers are "small" (if you pick a giant image for a scatter marker, too sad for you). 

> In the meantime, is there any other possible workaround?

There have been some proposals floated to afford more "manual" control over legends, e.g. 

* https://github.com/bokeh/bokeh/issues/10776

I am not sure I have any good immediate term suggestions. Or rather it depends. Do you need to draw 50 scatter markers or 50 thousand? Are they distinguished by color, so that a generic square of color in the legend could suffice? Do you know some javascript and are up for implementing a custom extensions? I'm happy to discuss some possibilities, but the Discourse would be a better venue to move that conversation: https://discourse.bokeh.org/
